File: meta.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <random>

using namespace std;

// ИНВАРИНТЫ:

//1-ое свойство: если буква b есть в строке, то при применении любого правила
//в строке всё равно будет хотя бы 1 буква b
bool f0(const string& prev, const string& curr){
    size_t found1 = prev.find("b");
    if (found1 != std::string::npos) {
        size_t found2 = curr.find("b");
        if (found2 != std::string::npos){
            return true;
        }
        return false;
    }
    return true;
}

//2-ое свойство: число букв b не увеличивается
bool f1(const string& prev, const string& curr){
    if (count(prev.begin(), prev.end(), 'b') >= count(curr.begin(), curr.end(), 'b')){
        return true;
    }
    return false;
}


struct Rule{
    string l;
    string r; 
    Rule(const string& left, const string& right) : l(left), r(right) {}
};


vector<Rule> allRulesT2;

string alphabet = "ab";


void initRulesT2(){
    allRulesT2.push_back(Rule("aaaa", "a"));
    allRulesT2.push_back(Rule("aaab", "b"));
    allRulesT2.push_back(Rule("aaaba", "babb"));
    allRulesT2.push_back(Rule("baba", "baab"));
    allRulesT2.push_back(Rule("bab", "baa"));
    allRulesT2.push_back(Rule("babbb", "babba"));
    allRulesT2.push_back(Rule("aaba", "baa"));
    allRulesT2.push_back(Rule("baa", "abb"));
    allRulesT2.push_back(Rule("aba", "bb"));
    allRulesT2.push_back(Rule("bb", "ba"));
}

// l - левая граница подстроки в s (индекс), которую нужно заменить по правилу
void executeRule(Rule rule, string& s, int l){
    //string& replace (size_t pos, size_t len, const string& str);
    int len = rule.l.length();
    s.replace(l, len, rule.r);
}

//проверим, можно ли применить правило из списка Rules
pair<bool, int> checkIsRuleMayBeApplied(Rule rule, const string& s){
    pair<bool, int> p;
    size_t found = s.find(rule.l);
    p.second = found;

    if (found != std::string::npos) {
        p.first = true;
        return p;
    }
    
    p.first = false;
    return p;
}


//генерация случайного номера правила в системе T2
// n - сколько всего правил
int getRandomRuleT2(int n) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<int> dist(0, n - 1);
    return dist(gen);
}

//генерация случайной строки, заданной длины n, состоящей из символов 
//фиксированного алфавита
string getRandomString(int n, const string& alphabet) {
    string s;
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<int> dist(0, alphabet.size() - 1);
    for (int i = 0; i < n; i++){
        s += alphabet[dist(gen)];
    }
    return s;
}
 

// генерируем случайные 7 правил и проверим, что при
//выполнении правил сохраняются инварианты
//если нет, то возвращаем прооблемную строку
//и возвращаем номер 1-ого проблемного правила
//( -1 , если всё верно)
//а также вернём номер свойства, которое не выполнилось
pair<pair<string, int>, int> checkString(string s, bool (*f)(const string&, const string&), int num){
    pair<pair<string, int>, int> ans;
    ans.first.first = "";
    ans.first.second = -1;
    int n = allRulesT2.size();
    for (int i = 0; i < 7; i++){
        int ruleNum = getRandomRuleT2(n);
        Rule rule = allRulesT2[ruleNum];
        pair<bool, int> p = checkIsRuleMayBeApplied(rule, s);
        if (p.first == true){
            string copy = s;
            executeRule(rule, s, p.second);
            if (f(copy, s)){
                continue;
            }else{
                ans.second = num;
            }

            ans.first.first = copy;
            ans.first.second = ruleNum;
            return ans;
        }
    }
    return ans;
}


//будем генерировать 50000 тестов
//вернём true, если все тесты успешны,
//иначе false
//и возвращаем строку, на которой была ошибка
//и номер правила, на котором инвариант не сохранился
//и номер инварианта
pair<bool, pair<string, pair<int, int>>> meta(bool (*f)(const string&, const string&), int num){
    pair<bool, pair<string, pair<int, int>>> p1;
    p1.first = true;
    p1.second.first = "";
    p1.second.second.first = -1;
    p1.second.second.second = -1;
    for (int i = 0; i < 50000; i++){
        string s = getRandomString(15, alphabet);
        pair<pair<string, int>, int> p2 = checkString(s, f, num);
        if (p2.first.second != -1){
            p1.first = false;
            p1.second.first = p2.first.first;
            p1.second.second.first = p2.first.second; 
            p1.second.second.second = p2.second;
            return p1;
        }
        
    }
    return p1;
}

int main(){
    initRulesT2();
    //вектор инвариантов
    vector<bool (*)(const string&, const string&)> invariants;
    invariants.push_back(&f0);
    invariants.push_back(&f1);
    
    //номер инварианта
    int num = 0;
    for (auto invariant: invariants){
        pair<bool, pair<string, pair<int, int>>> p = meta(invariant, num);
        cout <<"инвариант номер: " << num << endl;
        if (p.first){
            cout << " для этого инварианта все тесты успешны" << endl;
        }else{
            cout << " ошибка на строке: " << p.second.first << endl;
            cout << " проблемное правило: ";
            cout << allRulesT2[p.second.second.first].l << " -> " << allRulesT2[p.second.second.first].r << endl;
        }
        cout << endl;
        num += 1;
    }

}

File: meta.rs
use std::time::{SystemTime, UNIX_EPOCH};

struct SimpleRng {
    state: u64,
}

impl SimpleRng {
    fn new() -> Self {
        let seed = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos() as u64;
        Self { state: seed }
    }
    
    fn gen_range(&mut self, range: std::ops::Range<usize>) -> usize {
        self.state = self.state.wrapping_mul(6364136223846793005).wrapping_add(1);
        let start = range.start;
        let end = range.end;
        start + (self.state as usize) % (end - start)
    }
}

// ИНВАРИНТЫ:

//1-ое свойство: если буква b есть в строке, то при применении любого правила
//в строке всё равно будет хотя бы 1 буква b
fn f0(prev: &str, curr: &str) -> bool {
    let found1 = prev.find('b');
    if found1.is_some() {
        let found2 = curr.find('b');
        if found2.is_some() {
            return true;
        }
        return false;
    }
    true
}

//2-ое свойство: число букв b не увеличивается
fn f1(prev: &str, curr: &str) -> bool {
    let count_prev = prev.chars().filter(|&c| c == 'b').count();
    let count_curr = curr.chars().filter(|&c| c == 'b').count();
    count_prev >= count_curr
}

//3-ье свойство: Пусть матрица A = 1 0 и матрица B = 2 -1
//                                 0 1               2 -1

// Если заменить в строке все буквы a на матрицу A , а буквы b на матрицу B,
//то можно посчитать произведение. В исходной системе это произведение не будет меняться
// при любом правиле
 
type Matrix = [[i32; 2]; 2];

const A: Matrix = [[1, 0], [0, 1]];
const B: Matrix = [[2, -1], [2, -1]];

//перемножаем 2 матрицы
fn multiply_matrices(x: &Matrix, y: &Matrix) -> Matrix {
    let mut result = [[0; 2]; 2];
    for row in 0..2 {
        for column in 0..2 {
            for k in 0..2 {
                result[row][column] += x[row][k] * y[k][column];
            }
        }
    }
    result
}

//перемножаем все матрицы
fn multiply_all_matrices(s: &str) -> Matrix {
    if s.is_empty() {
        return [[1, 0], [0, 1]]; // единичная матрица
    }
    
    let mut chars = s.chars();
    let first_char = chars.next().unwrap();
    let mut result = if first_char == 'a' { A } else { B };
    
    for c in chars {
        let matrix = if c == 'a' { A } else { B };
        result = multiply_matrices(&result, &matrix);
    }
    
    result
}

fn print_matrix(a: &Matrix) {
    for row in 0..2 {
        for column in 0..2 {
            print!("{} ", a[row][column]);
        }
        println!();
    }
}

fn f2(prev: &str, curr: &str) -> bool {
    let a = multiply_all_matrices(prev);
    let b = multiply_all_matrices(curr);
    
    for row in 0..2 {
        for column in 0..2 {
            if a[row][column] != b[row][column] {
                return false;
            }
        }
    }
    true
}

struct Rule {
    l: String,
    r: String,
}

impl Rule {
    fn new(left: &str, right: &str) -> Self {
        Self {
            l: left.to_string(),
            r: right.to_string(),
        }
    }
}

static mut ALL_RULES_T2: Vec<Rule> = Vec::new();
const ALPHABET: &str = "ab";

fn init_rules_t2() {
    unsafe {
        ALL_RULES_T2.push(Rule::new("bb", "ba"));
        ALL_RULES_T2.push(Rule::new("aba", "ba"));
        ALL_RULES_T2.push(Rule::new("baa", "ba"));
        ALL_RULES_T2.push(Rule::new("bab", "ba"));
        ALL_RULES_T2.push(Rule::new("aaaa", "a"));
        ALL_RULES_T2.push(Rule::new("aaab", "b"));
    }
}

fn execute_rule(rule: &Rule, s: &mut String, l: usize) {
    let len = rule.l.len();
    s.replace_range(l..l + len, &rule.r);
}

fn check_is_rule_may_be_applied(rule: &Rule, s: &str) -> (bool, usize) {
    if let Some(found) = s.find(&rule.l) {
        (true, found)
    } else {
        (false, 0)
    }
}

fn get_random_rule_t2() -> usize {
    let mut rng = SimpleRng::new();
    unsafe {
        rng.gen_range(0..ALL_RULES_T2.len())
    }
}

fn get_random_string(n: usize) -> String {
    let mut rng = SimpleRng::new();
    let alphabet: Vec<char> = ALPHABET.chars().collect();
    (0..n)
        .map(|_| {
            let idx = rng.gen_range(0..alphabet.len());
            alphabet[idx]
        })
        .collect()
}

// генерируем случайные 7 правил и проверим, что при
//выполнении правил сохраняются инварианты
//если нет, то возвращаем прооблемную строку
//и возвращаем номер 1-ого проблемного правила
//( -1 , если всё верно)
//а также вернём номер свойства, которое не выполнилось
fn check_string(s: &str, f: fn(&str, &str) -> bool, num: i32) -> ((String, i32), i32) {
    let mut current = s.to_string();
    let mut ans = ((String::new(), -1), -1);
    
    unsafe {
        for _ in 0..7 {
            let rule_num = get_random_rule_t2();
            let rule = &ALL_RULES_T2[rule_num];
            let (can_apply, pos) = check_is_rule_may_be_applied(rule, &current);
            
            if can_apply {
                let copy = current.clone();
                execute_rule(rule, &mut current, pos);
                
                if !f(&copy, &current) {
                    ans.0 .0 = copy;
                    ans.0 .1 = rule_num as i32;
                    ans.1 = num;
                    return ans;
                }
            }
        }
    }
    
    ans
}

//будем генерировать 50000 тестов
//вернём true, если все тесты успешны,
//иначе false
//и возвращаем строку, на которой была ошибка
//и номер правила, на котором инвариант не сохранился
//и номер инварианта
fn meta(f: fn(&str, &str) -> bool, num: i32) -> (bool, (String, (i32, i32))) {
    for _ in 0..50000 {
        let s = get_random_string(15);
        let result = check_string(&s, f, num);
        
        if result.0 .1 != -1 {
            return (false, (result.0 .0, (result.0 .1, result.1)));
        }
    }
    
    (true, (String::new(), (-1, -1)))
}

fn main() {
    init_rules_t2();
    
    //вектор инвариантов
    let invariants: Vec<fn(&str, &str) -> bool> = vec![f0, f1, f2];
    
    //номер инварианта
    let mut num = 0;
    for invariant in invariants {
        let (success, (error_string, (rule_num, invariant_num))) = meta(invariant, num);
        println!("инвариант номер: {}", num);
        if success {
            println!(" для этого инварианта все тесты успешны");
        } else {
            println!(" ошибка на строке: {}", error_string);
            unsafe {
                if rule_num >= 0 {
                    let rule = &ALL_RULES_T2[rule_num as usize];
                    println!(" проблемное правило: {} -> {}", rule.l, rule.r);
                }
            }
        }
        println!();
        num += 1;
    }
}